File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
//이분탐색,매개변수탐색
int k, n; // 랜선의 개수, 만들어야하는 랜선의 개수
vector<int> lan;

void input()
{
	cin >> k >> n;
	lan.assign(k, 0);

	for (int i = 0; i < k; i++) {
		cin >> lan[i];
	}
}

bool isPossible(int len) {
	int cnt = 0;

	for (int i = 0; i < k; i++)
	{
		cnt += lan[i] / len;
	}
	if (cnt >= n) return true;
	else  return false;
}

void solve() {
	sort(lan.begin(), lan.end());

	unsigned int left = 1, right = lan[lan.size() - 1];
	unsigned int result = 0;

	while (left <= right) {
		unsigned int mid = (left + right) / 2;

		if (isPossible(mid)) {
			result = max(result, mid);
			left = mid + 1;
		}
		else {
			right = mid - 1;
		}
	}
	cout << result;
}
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	input();
	solve();
	return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

n, m = map(int,input().split())
lans = []

for _ in range(n):
    x = int(input())
    lans.append(x)

max_length = max(lans)

# 랜선을 x 단위로 자를 때 몇 개의 랜선을 얻을 수 있는가?
def lan_count(x):
    cnt = 0
    
    for l in lans:
        cnt += l // x
    return cnt

def binary_search():
    left = 1
    right = max_length
    res = 0

    while left <= right:
        mid = (left + right) // 2

        if lan_count(mid) < m:
            right = mid - 1
        elif lan_count(mid) >= m:
            left = mid + 1
            res = mid
    return res

print(binary_search())